odoo_test_helper/fake_model_loader.py

```diff
@@ -52,40 +52,38 @@ def tearDown(self):
             super().tearDown()
     """
 
-    _original_registry = None
-    _original_module2modules = None
-    _module_name = None
-
     def __init__(self, env, __module__):
         self.env = env
         self._module_name = self.env.registry["base"]._get_addon_name(__module__)
+        self._original_registry = None
+        self._new_models = []
 
     def backup_registry(self):
         self._original_registry = {}
-        self._original_module_to_models = {}
         for model_name, model in self.env.registry.models.items():
             self._original_registry[model_name] = {
                 "base": model.__bases__,
                 "_fields": model._fields.copy(),
                 "_inherit_children": OrderedSet(model._inherit_children._map.keys()),
                 "_inherits_children": set(model._inherits_children),
             }
-        for key in module_to_models:
-            self._original_module_to_models[key] = list(module_to_models[key])
 
     def _clean_module_to_model(self):
-        for key in self._original_module_to_models:
-            module_to_models[key] = list(self._original_module_to_models[key])
+        for module, models_ in module_to_models.items():
+            # filter out new models
+            module_to_models[module] = [x for x in models_ if x not in self._new_models]
 
     def update_registry(self, odoo_models):
         # Ensure that fake model are in your module
-        # If you test are re-using fake model form an other module
+        # If you test are re-using fake model from an other module
         # the following code will inject it like it was in your module
         self._clean_module_to_model()
 
         for model in odoo_models:
             if model not in module_to_models[self._module_name]:
                 module_to_models[self._module_name].append(model)
+            if model not in self._original_registry:
+                self._new_models.append(model)
```

Comment:
This is not correct.

When importing a python file, all odoo class of the file will be added into the module_to_models
https://github.com/odoo/odoo/blob/14.0/odoo/models.py#L171

So the new_models are not only the class passed, but all the class that exist in the file.

Moreover the variable "model" and the "_original_registry" do not contain the same thing.
In model it's a python class (that inherit of models.Models or models.Abstract...). In the registry it's a class builded by odoo (the final class with all stuff inherited)